### PR TITLE
Restore messages from s3

### PR DESCRIPTION
**Pull request by MacRudDNA**
…t of records

### CODE REVIEW

File: src/main/java/de/azapps/kafkabackup/common/partition/cloud/S3BatchDeserializer.java [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+
+    public List<Record> deserialize(InputStream inputStream) throws IOException {
+        byte[] batch = inputStream.readAllBytes();
+
+        List<Record> records = new ArrayList<>();
+
+        int lastRecordStart = 0;
+
+        for (int i = 0; i < batch.length; i++) {
+            if (batch[i] == UTF8_UNIX_LINE_FEED) {
```
Comment (loffek) [inline src/main/java/de/azapps/kafkabackup/common/partition/cloud/S3BatchDeserializer.java]: Hm. Doesn't jackson support reading ndjson? There seems to be some discussion here: https://github.com/FasterXML/jackson-future-ideas/issues/33
Suggesting something like https://github.com/FasterXML/jackson-future-ideas/issues/33#issuecomment-480242126. Perhaps worth investigating?

Comment (MacRudDNA) [reply src/main/java/de/azapps/kafkabackup/common/partition/cloud/S3BatchDeserializer.java]: Thanks for the tip :-)

File: src/main/java/de/azapps/kafkabackup/restore/message/PartitionMessageWriterWorker.java [hunk 2]
```diff
@@ -1,27 +1,127 @@
...
+
+  private void produceRecords(List<Record> recordsToProduce) {
+    try {
+      kafkaProducer.beginTransaction();
+      List<List<Record>> partitionedRecords = Lists.partition(recordsToProduce, PRODUCER_BATCH_SIZE);
+
+      for (List<Record> batch : partitionedRecords) {
+        List<Pair<Record, Future<RecordMetadata>>> futures = new ArrayList<>();
+        batch.forEach(record -> {
+          if (topicPartitionToRestore.getRestoredMessageInfoMap().containsKey(record.kafkaOffset())) {
```
Comment (loffek) [inline src/main/java/de/azapps/kafkabackup/restore/message/PartitionMessageWriterWorker.java]: Sorry, if I read this code too sloppily (it's getting late 😑) but it seems we are populating this map in the future, when the record has been written. Doesn't that mean that until a record is written, we don't have it in the map, and therefore we don't detect the duplicate?

File: src/main/java/de/azapps/kafkabackup/restore/message/RestoreMessageS3Service.java [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+
+
+  public List<String> getMessageBackupFileNames(String topicName, int partition) {
+    final String prefix = String.format("%s/%03d/", topicName, partition);
+
+    return awsS3Service.getBucketObjectKeys(bucketName, prefix, "/")
+        .stream()
+        .filter(name -> name.contains("msg_"))
+        .sorted(Comparator.naturalOrder())
+        .collect(Collectors.toList()); //TODO remove when offsets moved to separate bucket;
```
Comment (loffek) [inline src/main/java/de/azapps/kafkabackup/restore/message/RestoreMessageS3Service.java]: but keep the sorting! :D

**Review by loffek — COMMENTED** [review]
LGTM! But see my question around de-dup logic

File: src/main/java/de/azapps/kafkabackup/restore/message/RestoreMessageS3Service.java (hunk 3 above)
Comment (MacRudDNA) [reply src/main/java/de/azapps/kafkabackup/restore/message/RestoreMessageS3Service.java]: yup, just this last line :D

File: src/main/java/de/azapps/kafkabackup/restore/message/PartitionMessageWriterWorker.java (hunk 2 above)
Comment (MacRudDNA) [reply src/main/java/de/azapps/kafkabackup/restore/message/PartitionMessageWriterWorker.java]: Right, it worker when I was processing 1 by 1. Will fix it.

**Comment by MacRudDNA** [comment]
**How it works.**
Based on topics to restore allow and deny list we get list of TopicPartitionToRestore objects. For each there will be run one worker in thread pool with its own transactional producer.

Worker obtains list of json files with backup data and tries to send content of each batch in one transaction. After sending each message it's offset is saved in memory to prevent duplicates - before sending message we check if it's offset has already ben produced.
After committing transaction we read metadata of produced records to update new offset that each of the messages got.
So we end up with map of old-new offset for each restored message.

It is possible to force producer to use transaction for one produced record - to do this one needs to change RestoreMessageProducer.PRODUCER_BATCH_SIZE to **1**.

Please note that we produce each record with timestamp we have from backup. Looking at documentation [HERE](https://kafka.apache.org/documentation/#message.timestamp.type) I would argue that we can always give timestamp to produced message but for `CreateTime` type we can get some messages not being restored if restoring old messages. On the other hand if `message.timestamp.difference.max.ms` is same as `retention.ms` it should make no difference for us because older messages would also disappear from source cluster.

**Review by loffek — APPROVED** [review]
👨‍🍳

**Comment by loffek** [comment]
> Please note that we produce each record with timestamp we have from backup. Looking at documentation HERE I would argue that we can always give timestamp to produced message but for CreateTime type we can get some messages not being restored if restoring old messages. On the other hand if message.timestamp.difference.max.ms is same as retention.ms it should make no difference for us because older messages would also disappear from source cluster.

💯

**Pull request merged**